`Bad symbolic reference` when a java classfile refers to an annotation inner class not on the classpath

## Compiler version

Compiler version: 3.1.0-RC1


## Minimized code

```Scala
object Bot extends App {
  val token = sys.env("DISCORD_TEST_BOT")
  val gid = sys.env("GUILD_ID")
  val client: GatewayDiscordClient = (DiscordClient.create(token).login().block())
  val pingCmd = (ApplicationCommandRequest.builder()
                   .name ("ping")
                   .description ("Test ping command")
                   .build)

  val restClient = client.getRestClient
  val appId = restClient.getApplicationId.block

  (restClient
     .getApplicationService
     .createGuildApplicationCommand (appId, Snowflake.asLong(gid), pingCmd)
     .doOnError (err ⇒ println(s"Can't create command: ${err}"))
     .onErrorResume (e ⇒ Mono.empty())
     .block)

  (client
     on new ReactiveEventAdapter() {
       val cooldown_table = Map[Long, ZonedDateTime]()

       override def onSlashCommand(evt: SlashCommandEvent) = {
         if(evt.getCommandName() == "ping") {
         val reply = evt
           .getInteraction()
           .getMember()
           .toScala
           .map(mem ⇒ { "pong" })
           .getOrElse("Error: no member")
           evt.reply(reply)
         } else {
           Mono.empty()
         }
       }
     }
     blockLast)
}

```

## Output

```scala
[error] Bad symbolic reference. A signature
[error] refers to Value/T in package org.immutables.value which is not available.
[error] It may be completely missing from the current classpath, or the version on
[error] the classpath might be incompatible with the version used when compiling the signature.
[warn] Caught: java.lang.AssertionError: assertion failed: failure to resolve inner class:
[warn] externalName = org.immutables.value.Value$Immutable,
[warn] outerName = org.immutables.value.Value,
[warn] innerName = Immutable
[warn] owner.fullName = org.immutables.value.Value
[warn] while parsing ~/.cache/coursier/v1/https/oss.sonatype.org/content/repositories/snapshots/com/discord4j/discord4j-core/3.2.0-SNAPSHOT/discord4j-core-3.2.0-20210908.030540-190.jar(discord4j/core/spec/InteractionApplicationCommandCallbackReplyMonoGenerator.class) while parsing annotations in ~/.cache/coursier/v1/https/oss.sonatype.org/content/repositories/snapshots/com/discord4j/discord4j-core/3.2.0-SNAPSHOT/discord4j-core-3.2.0-20210908.030540-190.jar(discord4j/core/spec/InteractionApplicationCommandCallbackReplyMonoGenerator.class)
[warn] 5 warnings found
```

## Expectation

The code compiles (and run as expected) with the `Scala 2` compiler but give me this error message when compiling with the `Scala 3` compiler.

The full code and `sbt` build file can be found here: https://github.com/n3f4s/scala3D4j3.2ErrorMWE

A self-contained example:

**Value.java**

```java
public @interface Value {
    @interface Immutable {}
}

@Value.Immutable
abstract class Baz {}
```

**Test.scala**
```scala
object Test {
  def baz: Baz = ???
}
```

```
$ javac Value.java
$ rm Value*.class
$ scalac -3.head Test.scala
Bad symbolic reference. A signature
refers to Value/T in package <empty> which is not available.
It may be completely missing from the current classpath, or the version on
the classpath might be incompatible with the version used when compiling the signature.
Caught: java.lang.AssertionError: assertion failed: failure to resolve inner class:
externalName = Value$Immutable,
outerName = Value,
innerName = Immutable
owner.fullName = Value
while parsing ./Baz.class while parsing annotations in ./Baz.class
1 warning found
1 error found

$ scalac -213 Test.scala    // succeeds
```

--- a/compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala
+++ b/compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala
@@ -1127,7 +1127,16 @@ class ClassfileParser(
       val outerName = entry.strippedOuter
       val innerName = entry.originalName
       val owner = classNameToSymbol(outerName)
-      val result = atPhase(typerPhase)(getMember(owner, innerName.toTypeName))
+      val result = owner.denot.infoOrCompleter match
+        case _: StubInfo if hasAnnotation(entry.jflags) =>
+          requiredClass(innerName.toTypeName)
+            // It's okay for the classfiles of Java annotations to be missing
+            // from the classpath. If an annotation is defined as an inner class
+            // we need to avoid forcing the outer class symbol here, and instead
+            // return a new stub symbol for the inner class. This is tested by
+            // `surviveMissingInnerClassAnnot` in AnnotationsTests.scala
+        case _ =>
+          atPhase(typerPhase)(getMember(owner, innerName.toTypeName))
       assert(result ne NoSymbol,
         i"""failure to resolve inner class:
            |externalName = ${entry.externalName},

--- a/compiler/test/dotty/tools/AnnotationsTests.scala
+++ b/compiler/test/dotty/tools/AnnotationsTests.scala
@@ -66,3 +66,26 @@ class AnnotationsTest:
           s"A missing annotation while parsing a Java class should be silently ignored but: ${ctx.reporter.summary}")
       }
     }
+
+  @Test def surviveMissingInnerClassAnnot: Unit =
+    withJavaCompiled(
+      VirtualJavaSource("Outer.java",
+        """|package a.b;
+           |public @interface Outer { public @interface Value { @interface Immutable {} } }
+           |""".stripMargin),
+      VirtualJavaSource("Baz.java",
+        """|package a.b;
+           |@Outer.Value.Immutable abstract class Baz {}""".stripMargin)
+    ) { javaOutputDir =>
+      Files.delete(javaOutputDir.resolve("a/b/Outer.class"))
+      Files.delete(javaOutputDir.resolve("a/b/Outer$Value.class"))
+      Files.delete(javaOutputDir.resolve("a/b/Outer$Value$Immutable.class"))
+      inCompilerContext(javaOutputDir.toString + File.pathSeparator + TestConfiguration.basicClasspath) {
+        val cls = requiredClass("a.b.Baz")
+        val annots = cls.annotations.map(_.tree)
+        assert(annots == Nil,
+          s"class Baz should have no visible annotations since Outer.Value.Immutable is not on the classpath, but found: $annots")
+        assert(!ctx.reporter.hasErrors && !ctx.reporter.hasWarnings,
+          s"A missing annotation while parsing a Java class should be silently ignored but: ${ctx.reporter.summary}")
+      }
+    }

